orbotservice/src/main/java/org/torproject/android/service/OrbotService.java

```diff
@@ -916,142 +834,111 @@ private void updateLegacyV2OnionNames() throws SecurityException {
+                    if (Prefs.useDebugLogging()) {
+                        events.add(TorControlCommands.EVENT_DEBUG_MSG);
+                        events.add(TorControlCommands.EVENT_INFO_MSG);
+                    }
+                    conn.setEvents(events);
+                    logNotice("SUCCESS added control port event handler");
+                } catch (IOException | InterruptedException e) {
+                    e.printStackTrace();
+                    stopTorOnError(e.getLocalizedMessage());
+                    conn = null;
+                }
             }
 
-            if (exitCode != 0) {
-                logNotice("Tor did not start. Exit:" + exitCode);
-                return false;
+            @Override
+            public void onServiceDisconnected(ComponentName componentName) {
+                conn = null;
+                torService = null;
+                mEventHandler = null;
             }
 
-            //now try to connect
-            mLastProcessId = initControlConnection(10, false);
+            @Override
+            public void onNullBinding(ComponentName componentName) {
+                stopTorOnError("Tor was unable to start: " + "onNullBinding");
+                conn = null;
+                torService = null;
+                mEventHandler = null;
+            }
 
-            if (mLastProcessId == -1) {
-                logNotice(getString(R.string.couldn_t_start_tor_process_) + "; exit=" + exitCode);
-                throw new Exception(getString(R.string.couldn_t_start_tor_process_) + "; exit=" + exitCode);
-            } else {
-                logNotice("Tor started; process id=" + mLastProcessId);
+            @Override
+            public void onBindingDied(ComponentName componentName) {
+                stopTorOnError("Tor was unable to start: " + "onBindingDied");
+                conn = null;
+                torService = null;
+                mEventHandler = null;
             }
+        };
+
+        BroadcastReceiver receiver = new BroadcastReceiver() {
+            @Override
+            public void onReceive(Context context, Intent intent) {
+                if (TorService.ACTION_STATUS.equals(intent.getAction())
+                        && TorService.STATUS_ON.equals(intent.getStringExtra(TorService.EXTRA_STATUS))) {
+                    initControlConnection();
+                    unregisterReceiver(this);
+                }
+            }
+        };
+        // run the BroadcastReceiver in its own thread
+        HandlerThread handlerThread = new HandlerThread(receiver.getClass().getSimpleName());
+        handlerThread.start();
+        Looper looper = handlerThread.getLooper();
+        Handler handler = new Handler(looper);
+        registerReceiver(receiver, new IntentFilter(TorService.ACTION_STATUS), null, handler);
+
+        Intent serviceIntent = new Intent(this, TorService.class);
+        if (Build.VERSION.SDK_INT < 29) {
```

Comment:
Curious as to why this is handled differently on < 29

Reply:
because that version of `bindService()` was added in _android-29_:
https://developer.android.com/reference/android/content/Context#bindService(android.content.Intent,%20int,%20java.util.concurrent.Executor,%20android.content.ServiceConnection)

The new version of `bindService()` lets you specify the `Executor`.  Since there is already an `Executor` instance there, it can be reused on 29+